### PR TITLE
Store API Calls to get-sth

### PR DESCRIPTION
**Pull request by taknira**

### CODE REVIEW

File: monitor/api_call.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-monitor/client"
+)
+
+// APICall contains the details of a call to one of the API endpoints of a CT
+// Log.
+type APICall struct {
+	Start    time.Time
+	End      time.Time
```
[the remark this answers is not in the document]
Comment (taknira) [reply monitor/api_call.go]: As discussed in person, there's a good chance APICall.End will be needed to be explicitly stored so that all APICalls that occurred before a certain time can be obtained (as all events before a certain time translates to all APICalls that ended before a certain time).

However, as also discussed, it might be nice if this decision is left to the storage implementation - a Start and Duration can easily be converted into a Start and End.

Also as discussed, durations can be negative, so having a time.Duration here doesn't do anything to prevent silly things like API calls ending before they began, if, for example, something goes funny with the clock in between the two times being taken.

That being said, I want to think through whether there's anything outside of the storage implementation that might use this End value before making the change to Start and Duration.  Specifically, I suspect the availability calculating section of the monitor may need it, in which case I might prefer to keep it as Start and End.

In summary, I would like to keep this as Start and End for now, spend a small amount of time thinking about whether it is needed explicitly this way, and then, only if the answer to that is no, follow up with a PR to change End to Duration.

File: sthgetter/sthgetter.go [hunk 2]
```diff
@@ -7,27 +7,30 @@ import (
...
-func Run(ctx context.Context, lc *client.LogClient, url string, period time.Duration) {
+func Run(ctx context.Context, lc *client.LogClient, st storage.APICallWriter, url string, period time.Duration) {
 	log.Printf("%s: %s: Started with period %v", url, logStr, period)
 
 	t := time.NewTicker(period)
 	for {
 		select {
 		case <-t.C:
-			getCheckStoreSTH(url, lc)
+			getCheckStoreSTH(ctx, url, lc, st)
```
[the remark this answers is not in the document]
Comment (pphaneuf) [reply sthgetter/sthgetter.go]: I used to set timeouts at a lower level, for individual operations, which sounds a bit like what you're thinking, with timeouts possibly tailored to whatever the operation is (a Spanner read with a weak consistency ought to be done very quickly, but a write would have a longer timeout, for example)... But nowadays, my thinking is more toward a single high level timeout for the whole "request", and not fiddle with each sub-operation.

In particular, if you think about it, the question is this: is it ever okay for `getCheckStoreSTH` to never finish? Because that's what we're saying now, passing that indefinite context... Seems like something related to `period` might actually be what we want? After that, a sub-operation might use a tighter timeout, as well, but it wouldn't be strictly required.

Either way, I was only suggesting a TODO, this can be decided later, for this PR, go with this.

File: sthgetter/sthgetter.go [hunk 3]
```diff
@@ -7,27 +7,30 @@ import (
...
 // Run runs an STH Getter, which periodically gets an STH from a Log, checks
 // that each one meets per-STH requirements defined in RFC 6962, and stores
 // them.
-func Run(ctx context.Context, lc *client.LogClient, url string, period time.Duration) {
+func Run(ctx context.Context, lc *client.LogClient, st storage.APICallWriter, url string, period time.Duration) {
 	log.Printf("%s: %s: Started with period %v", url, logStr, period)
 
 	t := time.NewTicker(period)
 	for {
 		select {
```
Comment (pphaneuf) [inline sthgetter/sthgetter.go]: Maybe not in this PR, but I feel that this "infinite loop" that takes a context should abort if the context is done? Add the appropriate clause in the `select`? Optional for this PR, though, could be just a TODO, or a mental note... :smiley:

File: monitor/api_call.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pphaneuf) [reply monitor/api_call.go]: As we discussed, both ways have their pros and cons, neither is obviously better, so that sounds good to me.

**Comment by pphaneuf** [comment]
Done with this round of reviewing (I didn't click the right button, meant to do "start a review"!).

File: sthgetter/sthgetter.go (hunk 3 above)
Comment (taknira) [reply sthgetter/sthgetter.go]: Done (I think... check I did it right?)

File: sthgetter/sthgetter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (taknira) [reply sthgetter/sthgetter.go]: TODO added :)

File: sthgetter/sthgetter.go (hunk 3 above)
Comment (pphaneuf) [reply sthgetter/sthgetter.go]: I think so? :grin:
